[IDE] [Signature Help] Add basic signature help request to SourceKit

a7medev opened the pull request:
> [!NOTE]
> This PR depends on #83646, it should be its base PR but GitHub doesn't allow that.
> Commits after 6ebbb12 are part of this PR.

Add a basic `signaturehelp` request to SourceKit that:

1. Extracts and returns signature label.
2. Extracts offset and length of each parameter in the signature label.
3. Returns full documentation for signature (in a future PR, full documentation should be split into signature documentation and parameter documentation).

a7medev committed:
[IDE] Add basic signature help request support

[IDE] Refactor signature help to use argument completion directly

[IDE] Return signatures as out parameter of ArgumentTypeCheckCompletionCallback::getSignatures

a7medev committed:
[Test] Add signature help to sourcekitd-test

a7medev committed:
[IDE] Show signatures without a declaration

a7medev committed:
[Test] Add basic signature help label tests

a7medev committed:
NFC: Apply clang-format to signature help changes

a7medev committed:
[IDE] Resolve function type for curried function calls in ArgumentTypeCheckCompletionCallback

a7medev committed:
[IDE] Return whether a func is implicitly curried instance method in argument completion result

a7medev committed:
[IDE] Customize CodeCompletionStringBuilder for signature help

Print inout as a keyword rather than only &
Print default parameter value
Print _ in place of parameter label if not present

a7medev committed:
[IDE] Use CodeCompletionStringBuilder in signature help

a7medev committed:
[Test] Split signature help tests for currying

a7medev committed:
[IDE] NFC: Extract createSignatureLabel function & remove unused includes

a7medev committed:
[IDE] Output all parameters in signature help

a7medev committed:
[IDE] Use FunctionRefInfo::ApplyLevel for curried functions in ArgumentCompletion

And update signature help to output the signature correctly for a second apply and add tests in signature_help_currying

a7medev committed:
[IDE] NFC: Forward declare Signature in ArgumentCompletion.h

a7medev committed:
[IDE] Use raw documentation in signature help

a7medev committed:
[IDE] Use editor documents infrastructure for signature help

a7medev committed:
[IDE] Signature help: Keep strings in Allocator instead of SmallString

a7medev committed:
[IDE] Use keys.signatures instead of keys.members for signature help

a7medev committed:
[Test] Add signature help doc comment test

a7medev committed:
[Test] Change signature help tests to use split-file

hamishknight commented on lib/IDE/ArgumentCompletion.cpp:
```diff
@@ -95,14 +97,46 @@ bool ArgumentTypeCheckCompletionCallback::addPossibleParams(
+/// but not the second. We try to find the function's declaration and type
+/// if it turns out to be a double apply.
+static std::optional<std::pair<ValueDecl *, AnyFunctionType *>>
+tryResolveDoubleAppliedFunction(CallExpr *OuterCall, const Solution &S) {
```
`OuterCall` here can be null because the caller does a `dyn_cast`, you'll want to add an early return for that

a7medev replied:
Shouldn't `ParentCall` always be a `CallExpr`? I should probably just convert `dyn_cast` into `cast`. WDYT?

hamishknight replied:
Hmm, in practice I believe so, I think I would still rather be defensive here and `dyn_cast` though

a7medev commented on tools/SourceKit/lib/SwiftLang/SwiftSignatureHelp.cpp:
```diff
@@ -0,0 +1,272 @@
+  if (IsSecondApply) {
+    // Don't use the function decl when creating the signature label
+    // if this is the second apply of a double-applied function to
+    // avoid showing incorrect information like IUOs, rethrows, or
+    // reasync which can't be used in second apply.
```
TIL!

a7medev commented on lib/IDE/ArgumentCompletion.cpp:
```diff
@@ -95,14 +97,46 @@ bool ArgumentTypeCheckCompletionCallback::addPossibleParams(
+    if (!FuncRefInfo.isDoubleApply())
+      return std::nullopt;
+
+    auto CalleeTy = Overload->adjustedOpenedType->getAs<AnyFunctionType>();
+    auto ResultTy = S.simplifyTypeForCodeCompletion(CalleeTy->getResult());
+
+    if (auto *FuncTy = ResultTy->getAs<AnyFunctionType>())
+      return std::make_pair(Overload->choice.getDecl(), FuncTy);
```
This will work in the case of an implicitly curried instance method but not for a general second apply that's why I did it this way.

hamishknight replied:
Ah sorry, I missed that `ValueTy` is the type of the decl not the type of the call's function. This is fine

a7medev committed:
[IDE] Misc enhancements to signature help

a7medev commented on tools/SourceKit/include/SourceKit/Core/LangSupport.h:
```diff
@@ -1011,6 +1011,38 @@ class ConformingMethodListConsumer {
   virtual void cancelled() = 0;
 };
 
+struct SignatureHelpResult {
```
I see this pattern is already common (e.g. `ConformingMethodListResult`, `TypeContextInfoItem`, `CodeCompletionConsumer` have one in `IDE` and one in `SourceKit`). TBH, I agree that this is confusing but if I rename `SourceKit:: SignatureHelpResult` to `SourceKit:: SignatureHelpResponse` I'd prefer to rename things like `handleResult` to `handleResponse` which would drift a bit from existing conventions.

What do you think is best here?

hamishknight replied:
As discussed offline, we're going to leave this renamed

a7medev commented:
@hamishknight I've addressed your comments (half-addressed this one: https://github.com/swiftlang/swift/pull/83378#discussion_r2301420138), can you please recheck? 🙏🏼

hamishknight commented:
@a7medev Can you rebase this PR to make it a bit easier to re-review now that the other PRs have landed?

hamishknight commented on tools/SourceKit/lib/SwiftLang/SwiftSignatureHelp.cpp:
```diff
@@ -0,0 +1,270 @@
+                              : dyn_cast_or_null<AbstractFunctionDecl>(FD));
+  }
+
+  if (FD && FD->isImplicitlyUnwrappedOptional())
+    StringBuilder.addTypeAnnotationForImplicitlyUnwrappedOptional(
+        AFT->getResult(), DC, GenericSig);
+  else
+    StringBuilder.addTypeAnnotation(AFT->getResult(), DC, GenericSig);
```
Nit: Curly braces for `if-else`

```suggestion
  if (FD && FD->isImplicitlyUnwrappedOptional()) {
    StringBuilder.addTypeAnnotationForImplicitlyUnwrappedOptional(
        AFT->getResult(), DC, GenericSig);
  } else {
    StringBuilder.addTypeAnnotation(AFT->getResult(), DC, GenericSig);
  }
```

hamishknight reviewed (approved):
Thank you! LGTM

a7medev committed:
[IDE] NFC: Wrap if-else in curly braces

Co-authored-by: Hamish Knight <hamish_knight@apple.com>

hamishknight commented:
@swift-ci please smoke test

a7medev committed:
[Test] Avoid diff -B in signature help tests for Windows support

a7medev commented:
@hamishknight Fixed. I was mainly using it to prevent matching the extra newline at the end of the expected output but it wasn't necessary.

hamishknight commented:
Thanks!

hamishknight commented:
@swift-ci please smoke test

rintaro reviewed (commented):
I am fine with merging this as-is.

But my preference is that `tools/SourceKit` should only be a facade of `lib/IDE`.
I.e. all the logic should be in `IDE` and the primary test should be `swift-ide-test`.
`sourcekitd-test` is for testing SourceKit request/response handling/formatting only.

a7medev commented:
@rintaro I agree and we've considered this but we descoped it for now to focus on getting the feature done first. I plan on moving this to `lib/IDE` if I had time to work on it either during GSoC or after it.
